extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/scan/ScanQueryKit.java

```diff
@@ -159,14 +159,30 @@ public QueryDefinition makeQueryDefinition(
         clusterBy.getColumns()
     );
 
-    // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
-    // If there is a limit spec, it would write everything into one partition anyway, so the final shuffling should be
-    // applied after that.
+    ShuffleSpec scanShuffleSpec;
+    if (!hasLimitOrOffset) {
+      // If there is no limit spec, apply the final shuffling here itself. This will ensure partition sizes etc are respected.
+      scanShuffleSpec = finalShuffleSpec;
+    } else {
+      // If there is a limit spec, check if there are any non-boost columns to sort in.
+      boolean requiresSort = clusterByColumns.stream()
+                                             .anyMatch(keyColumn -> !QueryKitUtils.PARTITION_BOOST_COLUMN.equals(keyColumn.columnName()));
+      if (requiresSort) {
+        // If yes, do a sort into a single partition.
+        scanShuffleSpec = ShuffleSpecFactories.singlePartition().build(clusterBy, false);
+      } else {
+        // If the only clusterBy column is the boost column, we just use a mix shuffle to avoid unused shuffling.
+        // Note that we still need the boost column to be present in the row signature, since the limit stage would
+        // need it to be populated to do its own shuffling later.
```

Comment:
I was under the impression that the limit factory couldn't partition boost. Am I mistaken?

Reply:
otoh, if there isn't any partition boosting, then even the original code would have run into similar problem if there wasn't a cluster key right?

Reply:
The limit factory does not increment the partition boosting at regular intervals. The value would be all 0 if it was not added to the row signature at the scan stage.

> even the original code would have run into similar problem 

The original code would have mix shuffle speced. It would always make a single partition, so yes, it should have been too large.